pkg/api/v1/publish-api-types.go

```diff
@@ -0,0 +1,138 @@
+//    key uploads)
+//
+// Traveler - set to true if the TEKs in this publish set are consider to be the
+//  keys of a "traveler" who has left the home region represented by this server
+//  (or by the home health authority in case of a multi-tenant installation).
+//
+// RevisionToken: An opaque string that must be passed intact on additional
+//   publish requests from the same device, where the same TEKs may be published
+//   again.
+//
+// VisitedCountries: list (possibly empty) where the device has travelled to.
+//
+// ReportType: type of report - is it self-report, confirmed diagnose, ...?
+//
+// ConsentToFederation: whether user of the device has allowed to share his keys to other countries.
+//
+// Padding: random base64 encoded data to obscure the request size. The server will
+// not process this data in any way.
+//
+// Partial success: If at least one of the Keys passed in is valid, then the publish
+// request will accept those keys, return a response code of 200 (OK) AND also
+// return a 'Code' of ErrorPartialFailure along with an error message of
+// exactly which keys were not accepted and why. This does not indicate a failure
+// that must be reported to the user, but does indicate an issue with the application
+// making the upload (sending invalid data).
+type PublishKeysRequestDevice struct {
+	Keys                 []DeviceExposureKey `json:"temporaryExposureKeys" validate:"required"`
+	HealthAuthorityID    string              `json:"healthAuthorityID" validate:"required"`
+	VerificationPayload  string              `json:"verificationPayload"`
+	HMACKey              string              `json:"hmacKey"`
+	SymptomOnsetInterval int32               `json:"symptomOnsetInterval"`
+	Traveler             bool                `json:"traveler"`
+	RevisionToken        string              `json:"revisionToken" validate:"required"`
+
+	VisitedCountries    []string   `json:"visitedCountries"`
+	ReportType          ReportType `json:"reportType"`
+	ConsentToFederation bool       `json:"consentToFederation"`
+
+	Padding string `json:"padding"`
+}
+
+// PublishKeysResponseDevice is sent back to the client on a publish request.
+// If successful, the revisionToken indicates an opaque string that must be
+// passed back if the same devices wishes to publish TEKs again.
+//
+// On error, the error message will contain a message from the server
+// and the 'code' field will contain one of the constants defined in this file.
+// The intent is that code can be used to show a localized error message on the
+// device.
+//
+// The Padding field may be populated with random data on both success and
+// error responses.
+type PublishKeysResponseDevice struct {
+	RevisionToken     string `json:"revisionToken,omitempty"`
+	InsertedExposures int    `json:"insertedExposures,omitempty"`
+	ErrorMessage      string `json:"error,omitempty"`
+	Code              string `json:"code,omitempty"`
+	Padding           string `json:"padding,omitempty"`
+}
```

Comment:
this is 100% equal to upstream struct

Reply:
Yes. At this point. Are you sure there won't be a need for extension in the future? Or will we solve it then? :confused:

Reply:
it's upstream v1 api, they do a really good job by no introducing breaking changes to it and handle deprecation correctly (before there was v1alpha1 which they still support). It shouldn't happen then.

Reply:
No no, I meant what happens when I need to diverge the response to a device from the standard one. But It seems improbable when I think about that.

Reply:
well that should be handled the same, create a new api version, like `v2alpha1` and keep the old one for backward compatiblity :)